optuna/samplers/_search_space/group_decomposed.py

```diff
@@ -0,0 +1,103 @@
+import copy
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Tuple
+
+from optuna.distributions import BaseDistribution
+from optuna.study import BaseStudy
+from optuna.trial import TrialState
+
+
+class _SearchSpaceGroup(object):
+    def __init__(self) -> None:
+        self._search_spaces: List[Dict[str, BaseDistribution]] = []
+
+    @property
+    def search_spaces(self) -> List[Dict[str, BaseDistribution]]:
+        return self._search_spaces
+
+    def add_distributions(self, distributions: Dict[str, BaseDistribution]) -> None:
+        self._search_spaces = _add_distributions(self.search_spaces, distributions)
+
+
+def _add_distributions(
+    search_spaces: List[Dict[str, BaseDistribution]], distributions: Dict[str, BaseDistribution]
+) -> List[Dict[str, BaseDistribution]]:
+    if len(distributions) == 0:
+        return search_spaces
+
+    for search_space in search_spaces:
+        keys = set(search_space.keys())
+        dist_keys = set(distributions.keys())
+
+        if keys.isdisjoint(dist_keys):
+            continue
+
+        if keys < dist_keys:
+            return _add_distributions(
+                search_spaces, {name: distributions[name] for name in dist_keys - keys}
+            )
+
+        if keys > dist_keys:
+            search_spaces.append(distributions)
+            search_spaces.append({name: search_space[name] for name in keys - dist_keys})
+            search_spaces.remove(search_space)
+            return search_spaces
+
+        intersection = keys & dist_keys
+        search_spaces.append({name: search_space[name] for name in intersection})
+        if len(keys - intersection) > 0:
+            search_spaces.append({name: search_space[name] for name in keys - intersection})
+        search_spaces.remove(search_space)
+        return _add_distributions(
+            search_spaces, {name: distributions[name] for name in dist_keys - intersection}
+        )
```

Comment:
This logic can be more straightforward by removing the if statements.
```suggestion
        intersection = keys & dist_keys
        new_search_spaces = []
        new_search_spaces.append({name: search_space[name] for name in intersection})
        new_search_spaces.append({name: search_space[name] for name in keys - intersection})
        new_search_spaces.append({name: distributions[name] for name in dist_keys - intersection})
        search_spaces.remove(search_space)
        for new_search_space in new_search_spaces:
            search_spaces = _add_distributions(search_spaces, new_search_space)

        return search_spaces
```

Reply:
Or we can rewrite this function as non-recursive.

Reply:
Thanks for the suggestion. I'll work on the refactoring as a follow-up as we discussed privately.